mongodb: fixes build and sanitize package

bryanasdev000 opened the pull request:
###### Description of changes

This should close https://github.com/NixOS/nixpkgs/issues/171928 and https://github.com/NixOS/nixpkgs/issues/155121.

Basically a umbrella PR to fix MongoDB issues on Nixpkgs.

With help from @andersk @bachp @kfiz @superherointj @Artturin, thanks a lot!
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [X] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

bryanasdev000 committed:
mongodb: fix build and sanitize package

mongodb: fix build for 4.4

mongodb: fix build for 4.2

mongodb: fix build for 4.0

mongodb: 5.0.7 -> 5.0.8

mongodb: default to mongodb-5_0

mongodb: eol versions older than 4.0

mongodb: format and rename

mongodb: use gcc10 for 5.x

mongodb: use gcc10 only on linux

mongodb: update docs with eol and default version

mongodb: rename 5.0 to v5_0

mongodb: fix typo in doc

mongodb: change 5.x top level entry

mongodb: use stateversion to determine mongodb pkg

mongodb: use fetchpatch for gentoo patches

mongodb: drop local patch from gentoo

mongodb: rebase on master

Artturin commented on pkgs/servers/nosql/mongodb/v4_0.nix:
```diff
@@ -12,6 +12,6 @@ in buildMongoDB {
   sha256 = "sha256-ct33mnK4pszhYM4Is7j0GZQRyi8i8Qmy0wcklyq5LjM=";
   patches =
-    [ ./forget-build-dependencies.patch ./mozjs-45_fix-3-byte-opcode.patch ]
+    [ ./forget-build-dependencies.patch ./mozjs-45_fix-3-byte-opcode.patch ./mongodb-4.4.1-gcc11.patch ./mongodb-4.0-glibc-2.34.patch ] # https://github.com/NixOS/nixpkgs/issues/171928 https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-db/mongodb/files/mongodb-4.4.1-gcc11.patch https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-db/mongodb/files/mongodb-5.0.2-glibc-2.34.patch
```
Quite long now, please multiline

bryanasdev000 replied:
Done for all items.

bachp commented on pkgs/top-level/all-packages.nix:
```diff
@@ -22169,7 +22169,7 @@ with pkgs;
   mariadb = mariadb_106;
 
-  mongodb = hiPrio mongodb-3_4;
+  mongodb = hiPrio mongodb-5_0;
```
I think it's a good idea to document the default version change in a changelog entry.

bryanasdev000 replied:
Will do.

kfiz commented:
Please change this
`stdenv = gcc10Stdenv;`
in all-packages.nix 
to
`stdenv = if stdenv.cc.isGNU then gcc10Stdenv else stdenv;`
 or otherwise we run into problems on non-linux

kfiz commented:
> Folks, I think it's ready to ship, a squash is necessary?
> 
> I am just waiting for ofborg output before removing the draft.
> 
> CC: @andersk @bachp @kfiz @superherointj

LGTM. To nitpick, there is a minor typo in the doc sections: "provider" -> "provide".

bryanasdev000 commented:
> provider

Fixed, will probably need to squash :P

bryanasdev000 commented:
Hmm not sure why ofborg is complaining:

"This PR does not cleanly list package outputs after merging.".

bryanasdev000 commented:
@ofborg /eval

bryanasdev000 commented:
Well I think I can't force a new eval :P

bryanasdev000 commented:
@ofborg eval

thiagokokada commented:
@ofborg eval

bryanasdev000 commented:
Need to fix top level entry...

bryanasdev000 commented:
Ok this shoud work, thanks @superherointj !

Artturin commented:
https://github.com/NixOS/nixpkgs/blob/3c5ae9be1f18c790ea890ef8decbd0946c0b4c04/nixos/modules/services/databases/mongodb.nix#L34

https://search.nixos.org/options?channel=21.11&show=system.stateVersion&from=0&size=50&sort=relevance&type=packages&query=stateversion

Should probably be added to the module

bryanasdev000 commented:
> https://github.com/NixOS/nixpkgs/blob/3c5ae9be1f18c790ea890ef8decbd0946c0b4c04/nixos/modules/services/databases/mongodb.nix#L34
> 
> https://search.nixos.org/options?channel=21.11&show=system.stateVersion&from=0&size=50&sort=relevance&type=packages&query=stateversion
> 
> Should probably be added to the module

You mean a little if to check if state version is >= 22.05 we get mongodb else we get mongodb-3_6?

Artturin commented:
> > https://github.com/NixOS/nixpkgs/blob/3c5ae9be1f18c790ea890ef8decbd0946c0b4c04/nixos/modules/services/databases/mongodb.nix#L34
> > 
> > https://search.nixos.org/options?channel=21.11&show=system.stateVersion&from=0&size=50&sort=relevance&type=packages&query=stateversion
> > Should probably be added to the module
> 
> You mean a little if to check if state version is >= 22.05 we get mongodb else we get mongodb-3_6?

Yeah so then there's no silent upgrade, instead they'll see the insecure warning

bryanasdev000 commented:
> > > https://github.com/NixOS/nixpkgs/blob/3c5ae9be1f18c790ea890ef8decbd0946c0b4c04/nixos/modules/services/databases/mongodb.nix#L34
> > > 
> > > https://search.nixos.org/options?channel=21.11&show=system.stateVersion&from=0&size=50&sort=relevance&type=packages&query=stateversion
> > > Should probably be added to the module
> > 
> > 
> > You mean a little if to check if state version is >= 22.05 we get mongodb else we get mongodb-3_6?
> 
> Yeah so then there's no silent (multiversion) upgrade, instead they'll see the insecure warning

Ok, will check another module (postgres?) for inspiration :P 

Thanks.

bryanasdev000 commented:
> > > https://github.com/NixOS/nixpkgs/blob/3c5ae9be1f18c790ea890ef8decbd0946c0b4c04/nixos/modules/services/databases/mongodb.nix#L34
> > > 
> > > https://search.nixos.org/options?channel=21.11&show=system.stateVersion&from=0&size=50&sort=relevance&type=packages&query=stateversion
> > > Should probably be added to the module
> > 
> > 
> > You mean a little if to check if state version is >= 22.05 we get mongodb else we get mongodb-3_6?
> 
> Yeah so then there's no silent (multiversion) upgrade, instead they'll see the insecure warning

Done, based on MySQL and Postgres modules, will check how to test it locally now (or rely on of ofborg output if faster).

bryanasdev000 commented:
Well, ofborg was faster :P 

Trying to understand: "error: The option `cfg' does not exist. Definition values: ../nixos/modules/services/databases/mongodb.nix"

bryanasdev000 commented:
Ok, fixed! @Artturin could you please take a look? Thanks!

thiagokokada commented on pkgs/servers/nosql/mongodb/v4_4.nix:
```diff
@@ -10,5 +10,6 @@ buildMongoDB {
   sha256 = "sha256-ebg3R6P+tjRvizDzsl7mZzhTfqIaRJPfHBu0IfRvtS8=";
   patches = [
     ./forget-build-dependencies-4-4.patch
+    ./mongodb-4.4.1-gcc11.patch # https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-db/mongodb/files/mongodb-4.4.1-gcc11.patch
```
Here too.

bryanasdev000 replied:
Done.

bachp commented on pkgs/servers/nosql/mongodb/v4_0.nix:
```diff
@@ -11,7 +11,11 @@ let
+  patches = [
+    ./forget-build-dependencies.patch
+    ./mozjs-45_fix-3-byte-opcode.patch
+    ./mongodb-4.4.1-gcc11.patch # https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-db/mongodb/files/mongodb-4.4.1-gcc11.patch
```
Are we sure gentoo provides a stable URL? I don't see any hash or version in it. Might be more robust to just keep the copy here.

thiagokokada replied:
https://gitweb.gentoo.org/repo/gentoo.git/commit/dev-db/mongodb/files/mongodb-4.4.1-gcc11.patch?id=9bcbb4f7a076e7839c55678123c67a1c7b838237

Here you go.

bryanasdev000 replied:
I think there is a mirror on Github, but also vote to keep a copy here, another patch is a slighy modified version of ther patch, so we just organize what we need.

Also we probably would need to combo patches + fetchpatch :P

thiagokokada replied:
Either the GitHub mirror or this [URL](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-db/mongodb/files/mongodb-4.4.1-gcc11.patch?id=9bcbb4f7a076e7839c55678123c67a1c7b838237) should be fine.

> Also, we probably would need to combo patches + fetchpatch :package: 

I see no problem in this, this also happen in other packages.

thiagokokada replied:
Note, I am not completely against this approach, but using `fetchpatch` will make future bumps easier I think (instead of copying the contents of the patch in case of changes, we can just update the URL+hash). Also, it gives it more context.

bryanasdev000 replied:
Done.

bryanasdev000 commented:
@ofborg eval

superherointj commented:
@ofborg eval

Artturin commented:
@ofborg eval

bachp commented on pkgs/servers/nosql/mongodb/v4_0.nix:
```diff
@@ -11,7 +11,15 @@ let
+    ./mozjs-45_fix-3-byte-opcode.patch
+    ./mongodb-4.0-glibc-2.34.patch # https://github.com/NixOS/nixpkgs/issues/171928
+   (fetchpatch {
+      name = "mongodb-4.4.1-gcc11.patch";
```
Patch doesn't apply on 4.0, is there a reason we want to keep it around? My proposal is to just get rid of version 4.0.

bryanasdev000 replied:
Without the patch, it does not build, also I think the patch is more related to GCC11 than MongoDB version itself.

And yes, MongoDB 4.0 is already EOLed, but I think deprecating it now on Nixpkgs is not a good idea (it deprecated only last month).

andersk replied:
A reason to keep it around for the next NixOS release was discussed here: https://github.com/NixOS/nixpkgs/issues/155121#issuecomment-1120304376

bachp replied:
@andersk https://github.com/NixOS/nixpkgs/issues/155121#issuecomment-1120304376 is a good reason to keep it around.

@bryanasdev000 Sorry my mistake I missread the error.  It only fails on darwin and it is the `asio-no-experimental-string-view.patch` patch (darwin only) that fails to apply.
```
       last 10 log lines:
       > applying patch /nix/store/k0xlii3xyxlfparmx0w9ylmc75swgfbc-mongodb-4.0-glibc-2.34.patch
       > patching file src/mongo/stdx/thread.h
       > applying patch /nix/store/pmzplzw7nd78jrc3wf9y2jdarww58q6y-mongodb-4.4.1-gcc11.patch
       > patching file src/mongo/db/exec/plan_stats.h
       > Hunk #1 succeeded at 34 with fuzz 2 (offset 1 line).
       > applying patch /nix/store/f57czlnkrc71v6dh5k76msr3nc6jw12s-asio-no-experimental-string-view.patch
       > patching file src/third_party/asio-master/asio/include/asio/detail/config.hpp
       > Hunk #1 FAILED at 786.
       > Hunk #2 FAILED at 794.
       > 2 out of 2 hunks FAILED -- saving rejects to file src/third_party/asio-master/asio/include/asio/detail/config.hpp.rej

```

bachp commented:
Result of `nixpkgs-review pr 172009` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>mongodb (mongodb-5_0)</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>mongodb-4_0</li>
    <li>mongodb-4_2</li>
    <li>mongodb-4_4</li>
  </ul>
</details>

bryanasdev000 commented:
> Result of `nixpkgs-review pr 172009` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 
> 2 packages blacklisted:
> * nixos-install-tools
> * tests.nixos-functions.nixos-test
> 
> 1 package failed to build:
> * mongodb (mongodb-5_0)
> 
> 3 packages built:
> * mongodb-4_0
> * mongodb-4_2
> * mongodb-4_4
> 
> Details of the mongodb 5.0 error:
> 
> ```
> error: builder for '/nix/store/5iwjkqk6mm67g3z2vfv4d516gzwhvjq0-mongodb-5.0.8.drv' failed with exit code 2;
>        last 10 log lines:
>        > /nix/store/xbn6671zadppwl968j7x13bprvwiyssb-boost-1.77.0/lib/libboost_log.a(syslog_backend.o):function boost::asio::detail::scheduler::capture_current_exception(): error: undefined reference to 'std::__exception_ptr::exception_ptr::_M_release()'
>        > /nix/store/xbn6671zadppwl968j7x13bprvwiyssb-boost-1.77.0/lib/libboost_log.a(syslog_backend.o):function boost::asio::detail::scheduler::capture_current_exception(): error: undefined reference to 'std::__exception_ptr::exception_ptr::_M_release()'
>        > /nix/store/xbn6671zadppwl968j7x13bprvwiyssb-boost-1.77.0/lib/libboost_log.a(syslog_backend.o):function boost::asio::detail::scheduler::run(boost::system::error_code&): error: undefined reference to 'std::__exception_ptr::exception_ptr::_M_addref()'
>        > Compiling build/nixos/mongo/s/commands/cluster_move_chunk_cmd.o
>        > Compiling build/nixos/mongo/s/commands/cluster_cleanup_reshard_collection_cmd.o
>        > Compiling build/nixos/mongo/s/commands/cluster_drop_indexes_cmd.o
>        > collect2: error: ld returned 1 exit status
>        > scons: *** [build/nixos/mongo/db/mongod] Error 1
>        > scons: building terminated because of errors.
>        > build/nixos/mongo/db/mongod failed: Error 1
>        For full logs, run 'nix log /nix/store/5iwjkqk6mm67g3z2vfv4d516gzwhvjq0-mongodb-5.0.8.drv'.
> ```

Oh crap, did not expected v5.0 to break again, as soon as I get to my desktop I will try to debug it again.

Thanks for building :)

Et7f3 commented on nixos/modules/services/databases/mongodb.nix:
```diff
@@ -121,6 +120,15 @@ in
+     # ‘system.stateVersion’ to maintain compatibility with existing
+     # systems!
+     # Remember to change the default version at pkgs/top-level/all-packages.nix too between NixOS versions
+     mkDefault (if versionAtLeast config.system.stateVersion "22.05" then pkgs.mongodb-5_0
```
Do you need to put 22.11 since 22.05 is the default ?

bryanasdev000 replied:
Indeed, since we missed 22.05 release, tihs should be 22.11.

Since I only with a laptop now, I will delay a bit any change for now but I may give a look at weekend.

Et7f3 commented on pkgs/servers/nosql/mongodb/mongodb-4.0-glibc-2.34.patch:
```diff
@@ -0,0 +1,14 @@
+-    static constexpr std::size_t kMongoMinSignalStackSize = std::size_t{64} << 10;
+-
+-    static constexpr std::size_t kStackSize =
+-        std::max(kMongoMinSignalStackSize, std::size_t{MINSIGSTKSZ});
```
Do you prefer this upstream patch: https://github.com/mongodb/mongo/commit/ef08d0dbc99db8c4620512e92bfb3154282eb5d3 (just add .patch after url)

bryanasdev000 replied:
Sounds nice.

Et7f3 commented on pkgs/servers/nosql/mongodb/v4_4.nix:
```diff
@@ -10,5 +10,10 @@ buildMongoDB {
     ./forget-build-dependencies-4-4.patch
+   (fetchpatch {
+      name = "mongodb-4.4.1-gcc11.patch";
+      url = "https://raw.githubusercontent.com/gentoo/gentoo/7168257cad6ea7c4856b01c5703d0ed5b764367c/dev-db/mongodb/files/mongodb-4.4.1-gcc11.patch";
```
Do you prefer this commit from MongoDB dev ? [https://github.com/mongodb/mongo/commit/e78b2bf6eaa0c43bd76dbb841add167b443d2bb0](https://github.com/mongodb/mongo/commit/e78b2bf6eaa0c43bd76dbb841add167b443d2bb0.patch) same for 4.2 and 4.0

Et7f3 replied:
I did a bit of archeology to find those.

Et7f3 replied:
I used this patch in #178820 maybe you can cherry-pick my commit inside your branch.

bryanasdev000 replied:
Sounds nice too, if it builds it builds!

Did you fully built it and run with https://github.com/NixOS/nixpkgs/pull/178820? 

And just for the records, you used both https://github.com/mongodb/mongo/commit/107858d4d2ffd9f0690def5108898a8b4be70350 and https://github.com/mongodb/mongo/commit/e78b2bf6eaa0c43bd76dbb841add167b443d2bb0?

Et7f3 replied:
I added in flag "install-mongod" so it fully built only mongod and mongos. I might be mistaken but test where built but not executed. And I focused on 5.0 (I don't need to do migration) and only second patch was needed.

SuperSandro2000 replied:
Yes, please prefer fetching from upstream instead.

bryanasdev000 commented:
Well, with efforts from @superherointj and @Et7f3 sounds like we are closer than before to ship this PR!

I may give a shot to rebase tomorrow (It's late here now) and maybe someone can fully compile it to check end to end :P 

Thanks folks!

bryanasdev000 committed:
mongodb: update docs with eol and default version

bryanasdev000 committed:
mongodb: default to mongodb-5_0

bryanasdev000 committed:
mongodb: 5.0.8 -> 5.0.9

bryanasdev000 committed:
mongodb: backport openssl fix

bryanasdev000 committed:
mongodb: backport openssl fix to 4.4

bryanasdev000 commented:
So, I pushed a mess of tests, so far, MongoDB 4.4 and 5.0.9 still fails to build.

Will try it again when less sleepy, gist for logs https://gist.github.com/bryanasdev000/b086bbf9c9a0f278b9f6eb4b1670549e

bryanasdev000 committed:
mongodb: use default gcc version

bryanasdev000 commented:
Will try to build PR https://github.com/NixOS/nixpkgs/pull/178820 to check if it builds and what's the difference that's causing it to fail.

Et7f3 commented:
I tested on a Mac builder (with better CPU and more disk). It was also broken. It seem the issue was boost 1.79 you might be interested in cherry-pick https://github.com/Et7f3/nixpkgs/tree/fix-mongodb-darwin

bryanasdev000 commented:
> I tested on a Mac builder (with better CPU and more disk). It was also broken. It seem the issue was boost 1.79 you might be interested in cherry-pick #180620

Sorry for the delay @Et7f3, been busy with IRL.

I've tried to build 4.4, and I am leaving 5.0 in the queue to build on my laptop, but so far 4.4 fails with iostream errors on your branch and another full set of errors with cherry-picking (sadly did not catch the full log).

If mongo 4.4 and/or 5.0 is working for you on Darwin, I suggest that you unify your PRs (https://github.com/NixOS/nixpkgs/pull/178820, https://github.com/NixOS/nixpkgs/pull/178820 and any other necessary piece to make it fully work) and push it to review, so this PR will not be a blocker. Also, I can give some help in code review at least.

Note that I only have access to x86-64 architectures, so we may be chasing different errors :P

Et7f3 commented:
I don't think I will merge my two PR since it is targeted for 2 users classes. (Maybe their is overlap). From your logs it seem mongodb include special file only for linux hence it didn't caused issue on the darwin build.

Et7f3 commented:
Ok I did a round on linux and finally merged the 2 PR (since their is a overlap). See my PR for darwin and it should work for darwin and linux. (I am currently rebuilding after a rebase to be sure nothing got broken in the 2k commits)

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -22594,14 +22594,14 @@ with pkgs;
 
-  mongodb-5_0 = callPackage ../servers/nosql/mongodb/5.0.nix {
+  mongodb-5_0 = callPackage ../servers/nosql/mongodb/v5_0.nix {
     sasl = cyrus_sasl;
```
Do we even need this? It seems always to be cyrus_sasl.

kfiz replied:
Would it be clear which sasl library should be used if we removed it? There seem to be at least 2 options, namely cyrus_sasl and gsasl.

SuperSandro2000 replied:
Does mongodb even work with gsasl? If so this should rather be a boolean input to enable one or the other.

kfiz replied:
I think it would be possible to use gsasl:
https://www.mongodb.com/docs/manual/reference/program/mongos/#std-option-mongos.--ldapBindSaslMechanisms

kfiz commented:
> Ok I did a round on linux and finally merged the 2 PR (since their is a overlap). See my PR for darwin and it should work for darwin and linux. (I am currently rebuilding after a rebase to be sure nothing got broken in the 2k commits)

@Et7f3 @bryanasdev000 Should we try to wrap this up?

Et7f3 commented:
I tested my PR #180620 on both linux and mac 2 months ago and all builded. I didn't retested recently but it should still be in good shape. If you @kfiz mind reviewing. It is a full PR done.

bryanasdev000 commented:
I am away from my computer now, but I can try to dust it off and try again once I get back home.

We have 3 PRs now trying to fix Mongo at nixpkgs, this, @Et7f3 one and https://github.com/NixOS/nixpkgs/pull/190548 maybe now we can fix it for real.

bryanasdev000 commented:
> I tested my PR #180620 on both linux and mac 2 months ago and all builded. I didn't retested recently but it should still be in good shape. If you @kfiz mind reviewing. It is a full PR done. 

@Et7f3 since you have hardware for it, could you please retest atop of master? Just in case, to test Darwin specifics.

kfiz commented:
> I tested my PR #180620 on both linux and mac 2 months ago and all builded. I didn't retested recently but it should still be in good shape. If you @kfiz mind reviewing. It is a full PR done.

I tested the v5 patch on darwin and that worked. I'll have a look at the other adjustments later, but would give @bryanasdev000  the lead here as he has the best overview.

If we think we are good to go, we can consider updating the v5 build to 5.0.12 which is the latest patch release. I have a version of that here https://github.com/kfiz/nixpkgs/commit/b48aa952ab788e14e5af5e44055d19fc9716d185

SuperSandro2000 commented on nixos/modules/services/databases/mongodb.nix:
```diff
@@ -121,6 +120,15 @@ in
+     # ‘system.stateVersion’ to maintain compatibility with existing
+     # systems!
+     # Remember to change the default version at pkgs/top-level/all-packages.nix too between NixOS versions
+     mkDefault (if versionAtLeast config.system.stateVersion "22.11" then pkgs.mongodb-5_0
+           else if versionAtLeast config.system.stateVersion "22.05" then pkgs.mongodb-3_4
+           else throw "please set your desired package for state version ${config.system.stateVersion}");
```
```suggestion
     mkDefault (if versionAtLeast config.system.stateVersion "22.11" then pkgs.mongodb-5_0
           else pkgs.mongodb-3_4;
```

SuperSandro2000 commented on nixos/modules/services/databases/mongodb.nix:
```diff
@@ -32,8 +32,7 @@ in
       enable = mkEnableOption "the MongoDB server";
 
       package = mkOption {
-        default = pkgs.mongodb;
```
We should keep a default package

bryanasdev000 replied:
But this option will be overridden with the condition of stateVersion (a little bit below this snippet), so why define a version here? Also, if defined, won't conflict with this same condition?

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -92,6 +92,8 @@ In addition to numerous new and upgraded packages, this release has the followin
   as it requires `qt4`, which reached its end-of-life 2015 and will no longer be supported by nixpkgs.
   [According to Barco](https://www.barco.com/de/support/knowledge-base/4380-can-i-use-linux-os-with-clickshare-base-units) many of their base unit models can be used with Google Chrome and the Google Cast extension.
 
+- MongoDB's packages were updated to each latest major version and versions older than 4.0 marked with `meta.known Vulnerabilities` to flag as EOLed by upstream yet still in tree to provide a clean upgrade path, also MongoDB 5.x is the default version from now on.
```
```suggestion
- MongoDB's packages were updated to each latest major version and versions older than 4.0 marked with `meta.knownVulnerabilities` to flag as EOLed by upstream yet still in tree to provide a clean upgrade path, also MongoDB 5.x is the default version from now on.
```
that is not quite correct. ``meta.knownVulnerabilities`` has nothing to do with the maintenance status of upstream bit that the package contains long standing not fixed vulnerabilities which is the case because those packages are EOL.

bryanasdev000 replied:
This is what we found to mark EOL and give space to users to do a clean upgrade from 3.4 to the latest (be it 5.X or 6.X if https://github.com/NixOS/nixpkgs/pull/190548 gets merged first). 

There is another Nix mechanism to do that? 

And yes we need to keep all versions, as one should upgrade version by version until it reaches the desired one, see https://www.mongodb.com/docs/manual/release-notes/5.0-upgrade-standalone/ and https://github.com/NixOS/nixpkgs/issues/155121#issuecomment-1120298542 for reference.

Besides that, I think another way is doing the migration with a container then once the data is OK, start the service in NixOS with the desired version.

SuperSandro2000 commented on pkgs/servers/nosql/mongodb/mongodb.nix:
```diff
@@ -133,7 +133,7 @@ in stdenv.mkDerivation rec {
     homepage = "http://www.mongodb.org";
     inherit license;
-
+    knownVulnerabilities =  if (versionAtLeast version "4.0") then [] else [ "EOLed version, please check https://www.mongodb.com/support-policy/lifecycles and official docs on how to upgrade" ];
```
```suggestion
    knownVulnerabilities = lib.optionals (versionAtLeast version "4.0") [ "EOLed version, please check https://www.mongodb.com/support-policy/lifecycles and official docs on how to upgrade" ];
```
Please add concrete CVEs from https://www.mongodb.com/alerts#security into here. This field is not for marking EOL software without vulnerabilities.

bryanasdev000 replied:
Answered above.

bryanasdev000 commented:
Ok, thanks to @Et7f3 and @kfiz we now should have all mongodb versions in tree building (with exception being the v6.x per https://github.com/NixOS/nixpkgs/pull/190548#issuecomment-1242796859) in Darwin, thanks a lot!

Now, I think we should merge PR https://github.com/NixOS/nixpkgs/pull/190548 first and then focus on this one with all needed decisions, that include:

- Fix all builds on Linux
- Chose what, when and how to deprecate EOL Mongo versions
- Changelog
- Release update (if any)
- Refactors (if any)

This PR is also a little bit old and big, if needed we can split it in multiples, each one addressing some of the above points.

kfiz commented:
> This PR is also a little bit old and big, if needed we can split it in multiples, each one addressing some of the above points.
@SuperSandro2000 What would you suggest?

Et7f3 commented:
mongodb v6.0 landed nixpkgs so we can revive this PR

bryanasdev000 commented:
> mongodb v6.0 landed nixpkgs so we can revive this PR

And thanks to @kfiz we should have MongoDB 6.x running on Darwin land soon.

I will try to check what is missing for this in the weekend and redo this PR or split it in multiple.

dotlambda commented:
Can this be closed?

bryanasdev000 commented:
> Can this be closed?

I think we can, AFAIK we have both old and new versions working and mongodb 6 as default pkg.

What will be missing is to deprecate old versions (https://github.com/NixOS/nixpkgs/issues/195448) but honestly I am without energy to pursue this.

Thanks for the help everyone, if anyone need a bit of this PR, feel free to grab.
